src/backend/d3d12/CommandBufferD3D12.cpp

```diff
@@ -33,6 +36,146 @@ namespace d3d12 {
                     return DXGI_FORMAT_R32_UINT;
             }
         }
+
+        struct BindGroupStateTracker {
+            uint32_t cbvSrvUavDescriptorIndex = 0;
+            uint32_t samplerDescriptorIndex = 0;
+            DescriptorHeapHandle cbvSrvUavDescriptorHeap;
+            DescriptorHeapHandle samplerDescriptorHeap;
+            std::array<BindGroup*, kMaxBindGroups> bindGroups = {};
+            Device* device;
+
+            BindGroupStateTracker(Device* device) : device(device) {
+            }
+
+            void TrackSetBindGroup(const BindGroupLayoutBase* bindGroupLayout) {
+                const auto& layout = bindGroupLayout->GetBindingInfo();
+
+                for (size_t binding = 0; binding < layout.mask.size(); ++binding) {
+                    if (!layout.mask[binding]) {
+                        continue;
+                    }
+
+                    switch (layout.types[binding]) {
+                    case nxt::BindingType::UniformBuffer:
+                    case nxt::BindingType::StorageBuffer:
+                    case nxt::BindingType::SampledTexture:
+                        cbvSrvUavDescriptorIndex++;
+                    case nxt::BindingType::Sampler:
+                        samplerDescriptorIndex++;
+                    }
+                }
+            }
+
+            void SetBindGroup(Pipeline* pipeline, BindGroup* group, uint32_t index, ComPtr<ID3D12GraphicsCommandList> commandList) {
+                const auto& layout = group->GetLayout()->GetBindingInfo();
+
+                // these indices are the beginning of the descriptor table
+                uint32_t cbvSrvUavDescriptorStart = cbvSrvUavDescriptorIndex;
+                uint32_t samplerDescriptorStart = samplerDescriptorIndex;
+
+                bindGroups[index] = group;
+
+                PipelineLayout* pipelineLayout = ToBackend(pipeline->GetLayout());
+
+                // these indices are the offsets from the start of the descriptor table
+                uint32_t cbvIndex = pipelineLayout->GetDescriptorStartingIndex(index, PipelineLayout::Descriptor::Type::CBV);
+                uint32_t uavIndex = pipelineLayout->GetDescriptorStartingIndex(index, PipelineLayout::Descriptor::Type::UAV);
+                uint32_t srvIndex = pipelineLayout->GetDescriptorStartingIndex(index, PipelineLayout::Descriptor::Type::SRV);
+                uint32_t samplerIndex = pipelineLayout->GetDescriptorStartingIndex(index, PipelineLayout::Descriptor::Type::Sampler);
+
+                for (size_t binding = 0; binding < layout.mask.size(); ++binding) {
```

Comment:
Recording BindGroups in DescriptorSets is probably going to be one of the bottlenecks on this backend. We should try to pre-compute as much information as possible in d3d12::BindGroup, and fill the descriptor heap with a d3d12::BindGroup::RecordDescriptors(CPU_HANDLE heapStart)